### PR TITLE
Migrate kick, ban and unban command

### PR DESCRIPTION
**Pull request by RealYusufIsmail**
I have added a kick, ban and unban command. I will be adding more moderation commands in the future.
For the kick command you have to do `/kick @user` were as for the ban command you have to do `/ban @user del_day(int)`. For the unban command you will type `/unban user_id`

Partially implements #63 and #64 .

### CODE REVIEW

**Comment by RealYusufIsmail** [comment]
Unban/kick/ban commands work. The bot will be able to dm the user if there dms are not disabled.
Screenshot 2021-10-17 at 10 32 09

**Comment by RealYusufIsmail** [comment]
> Please allow the users to delete the messages for more than 7 days.

Fixed that.

**Comment by RealYusufIsmail** [comment]
I have fixed everything please can you check

**Comment by RealYusufIsmail** [comment]
I have not fixed everyting

**Comment by RealYusufIsmail** [comment]
> You still have some `TODO`s, do fix them.

which ones

**Comment by java-coding-prodigy** [comment]
Apparently Code-Factor is sending a false alarm. It says we have TODOs when we don't.

**Comment by RealYusufIsmail** [comment]
> Apparently Code-Factor is sending a false alarm. It says we have TODOs when we don't.

Oh ok.

**Comment by RealYusufIsmail** [comment]
@java-coding-prodigy Can you look at my latest change and see if should change it please.

**Comment by java-coding-prodigy** [comment]
Are you gonna be making a purge command now?

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target user is null");
+
+        Member targetMember = userOption.getAsMember();
+
+        Member author = Objects.requireNonNull(event.getMember(), "Author is null");
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:55]:
```suggestion
        Member author = Objects.requireNonNull(event.getMember(), "The author is null");
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            .addOption(OptionType.STRING, REASON_OPTION, "Why the user should be banned", true)
+            .addOptions(new OptionData(OptionType.INTEGER, DELETE_HISTORY_OPTION,
+                    "the amount of days of the message history to delete, none means no messages are deleted.",
+                    true).addChoice("none", 0).addChoice("recent", 1).addChoice("all", 7));
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target user is null");
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:51]:
```suggestion
                Objects.requireNonNull(event.getOption(USER_OPTION), "The target is null");
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+        if (!author.hasPermission(Permission.BAN_MEMBERS)) {
+            event.reply(
+                    "You can not unban users in this guild since you do not have the BAN_MEMBERS permission.")
+                .setEphemeral(true)
+                .queue();
+            return;
+        }
+
+        Member bot = Objects.requireNonNull(event.getGuild(), "The Bot is null").getSelfMember();
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java:55]:
```suggestion
        Member bot = Objects.requireNonNull(event.getGuild(), "The guild is null").getSelfMember();
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        Member targetMember =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The member is null")
+                    .getAsMember();
+
+        Member author = Objects.requireNonNull(event.getMember(), "Author is null");
+
+        String reason = Objects.requireNonNull(event.getOption(REASON_OPTION), "The reason is null")
+            .getAsString();
+
+        Member bot = Objects.requireNonNull(event.getGuild(), "The bot is null").getSelfMember();
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:52]:
```suggestion
        Member bot = Objects.requireNonNull(event.getGuild(), "The guild is null").getSelfMember();
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        if (!bot.hasPermission(Permission.BAN_MEMBERS)) {
+            event.reply(
+                    "I can not unban users in this guild since I do not have the BAN_MEMBERS permission.")
+                .setEphemeral(true)
+                .queue();
+
+            logger.error("The bot does not have BAN_MEMBERS permission on the server '{}' ",
+                    Objects.requireNonNull(event.getGuild()).getName());
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java:63]:
```suggestion
                    event.getGuild().getName());
```

**Review by Tais993 — CHANGES_REQUESTED** [review]

**Comment by RealYusufIsmail** [comment]
@Zabuzard Is there anything else i need to change?

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        getData().addOption(OptionType.USER, USER_OPTION, "The user who you want to kick", true)
+            .addOption(OptionType.STRING, REASON_OPTION, "Why the user should be kicked", true);
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target is null");
+
+        Member targetMember = userOption.getAsMember();
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:47]: Rename it to "target" to stay consistent

Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:47]: And I'd still check or the target is in this guild, if it's not in this guild it'd throw an error and that looks ugly for the logging

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+        if (!ModerationUtils.handleReason(reason, event)) {
+            return;
+        }
+
+        kickUser(targetMember, author, reason, event.getGuild(), event);
+    }
+
+    private static void kickUser(@NotNull Member member, @NotNull Member author,
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:97]:
```suggestion
    private static void kickUser(@NotNull Member target, @NotNull Member author,
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            .getAsString();
+
+        if (!ModerationUtils.handleReason(reason, event)) {
+            return;
+        }
+
+        unban(targetUser, reason, author, event);
+    }
+
+    private static void unban(@NotNull User targetUser, @NotNull String reason,
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java:77]:
```suggestion
    private static void unban(@NotNull User target, @NotNull String reason,
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    public UnbanCommand() {
+        super("unban", "Unbans a given user", SlashCommandVisibility.GUILD);
+
+        getData().addOption(OptionType.USER, USER_OPTION, "The user who you want to unban", true)
+            .addOption(OptionType.STRING, REASON_OPTION, "Why the user should be unbanned", true);
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        User targetUser = Objects.requireNonNull(event.getOption(USER_OPTION), "The user is null")
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java:42]:
```suggestion
        User target = Objects.requireNonNull(event.getOption(USER_OPTION), "The user is null")
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+     *        the {@link ModerationUtils#REASON_MAX_LENGTH} <br>
+     *        If the reason is under the limit it will pass and will allow the command to continue.
+     *
+     * @return {@link ModerationUtils#REASON_MAX_LENGTH}
+     */
+    public static boolean handleReason(@NotNull String reason, @NotNull SlashCommandEvent event) {
+        if (reason.length() <= REASON_MAX_LENGTH) {
+            return true;
+        }
+        event.reply("The reason can not be over " + REASON_MAX_LENGTH + " characters")
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java:30]:
```suggestion
        }
        
        event.reply("The reason can not be over " + REASON_MAX_LENGTH + " characters")
```

I'd myself add a line between for readability, just a suggestion, if you disagree you can keep it this way.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+                    "the amount of days of the message history to delete, none means no messages are deleted.",
+                    true).addChoice("none", 0).addChoice("recent", 1).addChoice("all", 7));
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target is null");
+
+        Member targetMember = userOption.getAsMember();
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:53]:
```suggestion
        Member target = userOption.getAsMember();
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java (hunk 6 above)
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:47]:
```suggestion
        Member target = userOption.getAsMember();
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java (hunk 7 above)
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:97]:
```suggestion
    private static void kickUser(@NotNull Member target, @NotNull Member author,
```

**Review by Tais993 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java (hunk 6 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:47]: No need since userOption checks

Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java]: How does it?
It accepts any user, so even that're not in this guild.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+                    + author.getUser().getAsTag() + " for: " + reason))
+            .queue();
+
+        logger.info(
+                " '{} ({})' banned the user '{} ({})' and deleted their message history of the last '{}' days. Reason being'{}'",
+                author.getUser().getAsTag(), author.getIdLong(), targetUser.getAsTag(),
+                targetUser.getIdLong(), deleteHistoryDays, reason);
+    }
+
+    private static boolean handleCanInteractWithTarget(@NotNull Member targetMember,
```
[the remark this answers is not in the document]
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:107]: @RealYusufIsmail

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java (hunk 6 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java]: no it gets the member then if the member is not null it gets it as a user

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
+package org.togetherjava.tjbot.commands.moderation;
+
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
+import org.jetbrains.annotations.NotNull;
+
+public enum ModerationUtils {
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java:8]: Missing javadoc.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * @param reason The Boolean reasonLimit will check if the reason is above the provided limit.
+     *        <br>
+     *        If it is it will throw an error and will tell the user that the reason can not be over
+     *        the {@link ModerationUtils#REASON_MAX_LENGTH} <br>
+     *        If the reason is under the limit it will pass and will allow the command to continue.
+     *
+     * @return {@link ModerationUtils#REASON_MAX_LENGTH}
+     */
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java:25]: * Missing description.
* Still bad javadoc in general.
* Incorrect return description

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
+        OptionMapping userOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target is null");
+
+        Member target = userOption.getAsMember();
+
+        Member author = Objects.requireNonNull(event.getMember(), "The author is null");
+
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:56]: Imo removing those newlines here would make it more readable.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        int deleteHistoryDays = Math
+            .toIntExact(Objects.requireNonNull(event.getOption(DELETE_HISTORY_OPTION)).getAsLong());
+
+        if (!ModerationUtils.handleReason(reason, event)) {
+            return;
+        }
+
+        banUser(userOption.getAsUser(), author, reason, deleteHistoryDays, event.getGuild(), event);
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:79]: You are still using `event.getGuild()` here, even though you introduced a `guild` variable beforehand...

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:106]: Still not done.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+public final class BanCommand extends SlashCommandAdapter {
+    private static final String USER_OPTION = "user";
+    private static final String DELETE_HISTORY_OPTION = "delete-history";
+    private static final String REASON_OPTION = "reason";
+    private static final Logger logger = LoggerFactory.getLogger(BanCommand.class);
+
+    /**
+     * Creates an instance of the ban command.
+     *
+     * @see BanCommand
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:36]: Whats this `@see BanCommand` thing to do here?

**Review by Zabuzard — CHANGES_REQUESTED** [review]

**Comment by RealYusufIsmail** [comment]
For Moderation utils for the reason check should I add a IllegalArgumentException @Zabuzard

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
+package org.togetherjava.tjbot.commands.moderation;
+
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.User;
+import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * This contains if statements that are common in all the moderation commands suck se the if
+ * statement for the reason length
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java:10]: Bad javadoc, english issues, reads super weird and is not on point. Rephrase:
> Utility class offering helpers revolving around user moderation, such as banning or kicking.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java (hunk 14 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java]: You only addressed the first bullet point.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+ * to ban the specific given user (for example a moderator attempting to ban an admin).
+ */
+public final class BanCommand extends SlashCommandAdapter {
+    private static final String USER_OPTION = "user";
+    private static final String DELETE_HISTORY_OPTION = "delete-history";
+    private static final String REASON_OPTION = "reason";
+    private static final Logger logger = LoggerFactory.getLogger(BanCommand.class);
+
+    /**
+     * Constructs an instance
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:34]: Missing dot.

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java (hunk 14 above)
Comment (RealYusufIsmail) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationUtils.java]: What do you mean

**Comment by Zabuzard** [comment]
Closing in favor of #244 .

**Pull request closed without merging**